ravenpy/models/importers.py

```diff
@@ -219,78 +219,82 @@ class RoutingProductGridWeightImporter:
         ), "Only one of gauge_ids or sub_ids can be specified"
 
-        if Path(shapefile_path).suffix == ".zip":
-            shapefile_path = f"zip://{shapefile_path}"
-        self._shapes = geopandas.read_file(shapefile_path)
-        # Note that we cannot use xarray because it complains about variables and dimensions
-        # having the same name.
-        self._data = nc4.Dataset(nc_data_path)
+        self._input_is_netcdf = True
+
+        input_file_path = Path(input_file_path)
+        if input_file_path.suffix == ".nc":
+            # Note that we cannot use xarray because it complains about variables and dimensions
+            # having the same name.
+            self._input_data = nc4.Dataset(input_file_path)
+        elif input_file_path.suffix == ".zip":
+            self._input_data = geopandas.read_file(f"zip://{input_file_path}")
+            self._input_is_netcdf = False
+        elif input_file_path.suffix == ".shp":
+            self._input_data = geopandas.read_file(input_file_path)
+            self._input_is_netcdf = False
+        else:
+            raise ValueError(
+                "The input file must be a shapefile (.shp or .zip) or NetCDF"
+            )
+
+        routing_file_path = Path(routing_file_path)
+        if routing_file_path.suffix == ".zip":
+            routing_file_path = f"zip://{routing_file_path}"
+        self._routing_data = geopandas.read_file(routing_file_path)
 
     def extract(self) -> GridWeightsCommand:
-        # Raven numbering is:
-        #
-        #      [      1      2      3   ...     1*nlon
-        #        nlon+1 nlon+2 nlon+3   ...     2*nlon
-        #           ...    ...    ...   ...     ...
-        #           ...    ...    ...   ...  nlat*nlon ]
-        #
-        # --> Making sure shape of lat/lon fields is like that
-        #
-        lon_var = self._data.variables[self._var_names[0]][:]
-        if self._data.variables[self._var_names[0]].dimensions == self._dim_names:
-            lon_var = np.transpose(lon_var)
-
-        lat_var = self._data.variables[self._var_names[1]][:]
-        if self._data.variables[self._var_names[1]].dimensions == self._dim_names:
-            lat_var = np.transpose(lat_var)
-
-        lath, lonh = self._create_gridcells_from_centers(lat_var, lon_var)
-
-        nlon = np.shape(lon_var)[1]
-        nlat = np.shape(lat_var)[0]
-
-        # shape     = shape.to_crs(epsg=crs_lldeg)        # this is lat/lon in degree
+
+        self._prepare_input_data()
+
+        # Read routing data
+
         # WGS 84 / North Pole LAEA Canada
-        self._shapes = self._shapes.to_crs(
+        self._routing_data = self._routing_data.to_crs(
             epsg=RoutingProductGridWeightImporter.CRS_CAEA
         )
 
-        self._shapes = self._shapes.drop_duplicates(self._hru_id_field).sort_values(
-            self._hru_id_field
-        )
+        self._routing_data = self._routing_data.drop_duplicates(
```

Comment:
I am unsure about dropping duplicates here...

Reply:
I am suggesting an additional test below...

Reply:
You seem to be doing it in your own code though: https://github.com/julemai/GridWeightsGenerator/blob/main/derive_grid_weights.py#L453-L463

This is just the idiomatic numpy way of doing it.

Reply:
yes, but _after_ I have build the tree...

Reply:
ok so you think we should tolerate ROUTING_SHAPEFILE rows having the same HRU_ID?

Reply:
Yes, it's a long story. There is nothing to resolve this for now. I hate this too but...